binaries/cuprated/src/logging.rs

```diff
@@ -0,0 +1,137 @@
+use std::{
+    fmt::{Display, Formatter},
+    mem::forget,
+    sync::OnceLock,
+};
+
+use tracing::{
+    instrument::WithSubscriber, level_filters::LevelFilter, subscriber::Interest, Metadata,
+};
+use tracing_appender::{non_blocking::NonBlocking, rolling::Rotation};
+use tracing_subscriber::{
+    filter::Filtered,
+    fmt::{
+        self,
+        format::{DefaultFields, Format},
+        Layer as FmtLayer,
+    },
+    layer::{Context, Filter, Layered, SubscriberExt},
+    reload::{Handle, Layer as ReloadLayer},
+    util::SubscriberInitExt,
+    Layer, Registry,
+};
+
+use cuprate_helper::fs::logs_path;
+
+use crate::config::Config;
+
+/// A [`OnceLock`] which holds the [`Handle`] to update the file logging output.
+///
+/// Initialized in [`init_logging`].
+static FILE_WRITER_FILTER_HANDLE: OnceLock<Handle<CupratedTracingFilter, Registry>> =
+    OnceLock::new();
+
+/// A [`OnceLock`] which holds the [`Handle`] to update the stdout logging output.
+///
+/// Initialized in [`init_logging`].
+#[expect(clippy::type_complexity)] // factoring out isn't going to help readability.
+static STDOUT_FILTER_HANDLE: OnceLock<
+    Handle<
+        CupratedTracingFilter,
+        Layered<
+            Filtered<
+                FmtLayer<Registry, DefaultFields, Format, NonBlocking>,
+                ReloadLayer<CupratedTracingFilter, Registry>,
+                Registry,
+            >,
+            Registry,
+            Registry,
+        >,
+    >,
+> = OnceLock::new();
+
+/// The [`Filter`] used to alter cuprated's log output.
+pub struct CupratedTracingFilter {
+    pub level: LevelFilter,
+}
+
+impl Display for CupratedTracingFilter {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Filter")
+            .field("minimum_level", &self.level.to_string())
+            .finish()
+    }
+}
+
+impl<S> Filter<S> for CupratedTracingFilter {
+    fn enabled(&self, meta: &Metadata<'_>, cx: &Context<'_, S>) -> bool {
+        Filter::<S>::enabled(&self.level, meta, cx)
+    }
+
+    fn callsite_enabled(&self, meta: &'static Metadata<'static>) -> Interest {
+        Filter::<S>::callsite_enabled(&self.level, meta)
+    }
+
+    fn max_level_hint(&self) -> Option<LevelFilter> {
+        Some(self.level)
+    }
+}
+
+/// Initialize [`tracing`] for logging to stdout and to a file.
+pub fn init_logging(config: &Config) {
+    // initialize the stdout filter, set `STDOUT_FILTER_HANDLE` and create the layer.
+    let (stdout_filter, stdout_handle) = ReloadLayer::new(CupratedTracingFilter {
+        level: config.tracing.stdout.level,
+    });
+
+    drop(STDOUT_FILTER_HANDLE.set(stdout_handle));
```

Comment:
Why not `unwrap` here to assert it was only set once? `CupratedTracingFilter` needs `#[derive(Debug)]` for this.
```suggestion
    STDOUT_FILTER_HANDLE.set(stdout_handle).unwrap();
```